shell/platform/windows/flutter_windows_texture_registrar.cc

```diff
@@ -78,19 +79,22 @@ int64_t FlutterWindowsTextureRegistrar::EmplaceTexture(
 }
 
 bool FlutterWindowsTextureRegistrar::UnregisterTexture(int64_t texture_id) {
-  {
+  engine_->task_runner()->RunNowOrPostTask([engine = engine_, texture_id]() {
+    engine->UnregisterExternalTexture(texture_id);
+  });
+
+  std::promise<bool> promise;
+  engine_->PostRasterThreadTask([this, &promise, texture_id]() {
     std::lock_guard<std::mutex> lock(map_mutex_);
     auto it = textures_.find(texture_id);
-    if (it == textures_.end()) {
-      return false;
+    if (it != textures_.end()) {
+      textures_.erase(it);
+      promise.set_value(true);
+    } else {
+      promise.set_value(false);
     }
-    textures_.erase(it);
-  }
-
-  engine_->task_runner()->RunNowOrPostTask([engine = engine_, texture_id]() {
-    engine->UnregisterExternalTexture(texture_id);
   });
-  return true;
+  return promise.get_future().get();
```

Comment:
@stuartmorgan @chinmaygarde the issue this is attempting to resolve is a race condition due to the fact that registering/unregistering textures via the embedder API delegates work to the raster thread, however, we're removing the backing texture immediately here, so any pending tasks on the raster thread that may still reference the texture can blow up.

It's a bit unfortunate that our public API includes a return value -- in particular one whose semantics differ from the [return value from the embedder API](https://github.com/flutter/engine/blob/b0b1f240555c92100583c6693a1d04d6456a777d/shell/platform/embedder/embedder.cc#L2076). I'm really not a fan of blocking here given that most of the time the return value doesn't really matter much. I kind of wonder if we can get away with altering the semantics ever-so-slightly to be a closer match to the embedder API's semantics given that [we don't currently document](https://github.com/flutter/engine/blob/b0b1f240555c92100583c6693a1d04d6456a777d/shell/platform/common/client_wrapper/include/flutter/texture_registrar.h#L100) what the return value actually means. @stuartmorgan probably has thoughts.

I wonder if a better approach to blowing away the texture here might be for us to tweak `FlutterOpenGLTexture` et al to support an `unregister_callback` that we'd call from the raster thread. Looks like we actually [already do](https://github.com/flutter/engine/blob/b0b1f240555c92100583c6693a1d04d6456a777d/common/graphics/texture.cc#L27) make this callback but it's currently implemented as [a no-op](https://github.com/flutter/engine/blob/b0b1f240555c92100583c6693a1d04d6456a777d/shell/platform/embedder/embedder_external_texture_gl.cc#L114).

That would also avoid the need to expose `PostRasterThreadTask` on the Windows `FlutterEngine` for now, as a nice side-effect.

Reply:
> It's a bit unfortunate that our public API includes a return value -- in particular one whose semantics differ from the [return value from the embedder API](https://github.com/flutter/engine/blob/b0b1f240555c92100583c6693a1d04d6456a777d/shell/platform/embedder/embedder.cc#L2076). I'm really not a fan of blocking here given that most of the time the return value doesn't really matter much. I kind of wonder if we can get away with altering the semantics ever-so-slightly to be a closer match to the embedder API's semantics given that [we don't currently document](https://github.com/flutter/engine/blob/b0b1f240555c92100583c6693a1d04d6456a777d/shell/platform/common/client_wrapper/include/flutter/texture_registrar.h#L100) what the return value actually means. @stuartmorgan probably has thoughts.
> 

The return value doesn't matter much in fact, however, I made this block to let the caller know when it's safe to deallocate corresponding buffers.

Reply:
> I made this block to let the caller know when it's safe to deallocate corresponding buffers.

I wonder if we could solve that by offering an overload of UnregisterTexture that supports a callback that would notify the user that we're completely done with the texture.

Reply:
Capturing from a separate discussion, we should make a new unregister method with a completion callback, and deprecate this unregister method, so that we're not adding blocking of a critical thread in order to handle buffer cleanup.

Reply:
Thanks for updating -- I'm out on holiday for a couple weeks, and in the mad rush to get packed (and break a toe in the process), didn't get back with an update here after our chat. 

@jnschulze if you wanted to send a patch adding an alternative unregister method with the completion callback, and deprecate the existing method, I'd be happy to review; otherwise I'm more than happy to send a patch when I'm back.

Reply:
@cbracken Yes, I want to add an alternative method as suggested, but just came back from vacation and haven’t had a chance yet :) Enjoy your holidays!

Reply:
@cbracken
I've finally had a chance to get back to this and added a new `flutter::TextureRegistrar::UnregisterTexture` overload accepting a callback.

I'm not sure how to deal with the old `bool TextureRegistrar::UnregisterTexture` though.
Options include
- Removing it (the Windows impl was broken anyway)
- Making it always return `true` without blocking (the approach I chose)
- Synchronously waiting for the callback to complete

Reply:
@cbracken @loic-sharma
Any thoughts on how to deal with the old overload?

Reply:
I agree with your approach of keeping the deprecated API but making it always return `true` without blocking. There may be apps that use these APIs, we should avoid breaking them if possible.